src/metabase/driver/sql_jdbc/execute.clj

```diff
@@ -259,10 +305,42 @@
           (.cancel stmt))))
     stmt))
 
-(defmethod execute-query! :sql-jdbc
+(defn- use-statement? [driver params]
+  (and (statement-supported? driver) (empty? params)))
+
+(defn- statement*
+  ^Statement [driver conn canceled-chan]
+  ;; if canceled-chan gets a message, cancel the Statement
+  (let [^Statement stmt (statement driver conn)]
+    (a/go
+      (when (a/<! canceled-chan)
+        (log/debug (trs "Query canceled, calling Statement.cancel()"))
+        (u/ignore-exceptions
+         (.cancel stmt))))
+    stmt))
+
+(defn- ^Statement statement-or-prepared-statement [driver conn sql params canceled-chan]
+  (if (use-statement? driver params)
+    (statement* driver conn canceled-chan)
+    (prepared-statement* driver conn sql params canceled-chan)))
+
+(defmethod ^ResultSet execute-prepared-statement! :sql-jdbc
   [_ ^PreparedStatement stmt]
   (.executeQuery stmt))
 
+(defmethod ^ResultSet execute-statement! :sql-jdbc
+  [driver ^Statement stmt ^String sql]
+  (if (.execute stmt sql)
+    (.getResultSet stmt)
+    (throw (ex-info (str (tru "Select statement did not produce a ResultSet for native query"))
+                    {:sql sql :driver driver}))))
+
+(defn- ^ResultSet execute-statement-or-prepared-statement [driver ^Statement stmt max-rows params sql]
```

Comment:
just noticed this. This function should end in a `!`, because it has side-effects